### PR TITLE
[SPARK-18294][CORE] Implement commit protocol to support `mapred` package's committer

### PR DESCRIPTION
**Pull request by jiangxb1987**
## What changes were proposed in this pull request?

This PR makes the following changes:

- Implement a new commit protocol `HadoopMapRedCommitProtocol` which support the old `mapred` package's committer;
- Refactor SparkHadoopWriter and SparkHadoopMapReduceWriter, now they are combined together, thus we can support write through both mapred and mapreduce API by the new SparkHadoopWriter, a lot of duplicated codes are removed.

After this change, it should be pretty easy for us to support the committer from both the new and the old hadoop API at high level.

## How was this patch tested?
No major behavior change, passed the existing test cases.

### CODE REVIEW

File: core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala [hunk 1]
```diff
@@ -561,7 +561,7 @@ class PairRDDFunctionsSuite extends SparkFunSuite with SharedSparkContext {
       pairs.saveAsHadoopFile(
         "ignored", pairs.keyClass, pairs.valueClass, classOf[FakeFormatWithCallback], conf)
     }
-    assert(e.getMessage contains "failed to write")
+    assert(e.getCause.getMessage contains "failed to write")
```
Comment (jiangxb1987) [inline core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala]: This was handled by `SparkHadoopWriter`, so we should fetch the error message from the Throwable.

**Comment by jiangxb1987** [comment]
retest this please

**Comment by jiangxb1987** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**